pkg/openstack/loadbalancer_test.go

```diff
@@ -870,3 +871,151 @@ func Test_getSecurityGroupName(t *testing.T) {
+				},
+			},
+			want: "annotation-test-id",
+		},
+		{
+			name: "test get subnet from config class",
+			args: args{
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							LBClasses: map[string]*LBClass{
+								"test-class": {
+									SubnetID:           "test-class-subnet-id",
+									FloatingNetworkID:  "test-floating-network-id",
+									FloatingSubnetID:   "test-floating-subnet-id",
+									FloatingSubnet:     "test-floating-subnet",
+									NetworkID:          "test-network-id",
+									FloatingSubnetTags: "test-floating-subnet-tags",
+									MemberSubnetID:     "test-member-subnet-id",
+								},
+							},
+						},
+					},
+				},
+				service: &corev1.Service{
+					ObjectMeta: v1.ObjectMeta{
+						Annotations: map[string]string{
+							"loadbalancer.openstack.org/class": "test-class",
+						},
+					},
+				},
+			},
+			want: "test-class-subnet-id",
+		},
+		{
+			name: "test get subnet from config class with invalid loadbalancer class",
+			args: args{
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							LBClasses: map[string]*LBClass{
+								"decoy-class": {
+									SubnetID:           "test-id",
+									FloatingNetworkID:  "test-floating-network-id",
+									FloatingSubnetID:   "test-floating-subnet-id",
+									FloatingSubnet:     "test-floating-subnet",
+									NetworkID:          "test-network-id",
+									FloatingSubnetTags: "test-floating-subnet-tags",
+									MemberSubnetID:     "test-member-subnet-id",
+								},
+							},
+							SubnetID: "test-subnet-id",
+						},
+					},
+				},
+				service: &corev1.Service{
+					ObjectMeta: v1.ObjectMeta{
+						Annotations: map[string]string{
+							"loadbalancer.openstack.org/class": "test-class",
+						},
+					},
+				},
+			},
+			want:        "",
+			expectedErr: fmt.Sprintf("invalid loadbalancer class %q", "test-class"),
+		},
+		{
+			name: "test get subnet from default config",
+			args: args{
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							LBClasses: map[string]*LBClass{
+								"test-config-class-subnet-id": {
+									SubnetID:           "test-id",
+									FloatingNetworkID:  "test-floating-network-id",
+									FloatingSubnetID:   "test-floating-subnet-id",
+									FloatingSubnet:     "test-floating-subnet",
+									NetworkID:          "test-network-id",
+									FloatingSubnetTags: "test-floating-subnet-tags",
+									MemberSubnetID:     "test-member-subnet-id",
+								},
+							},
+							SubnetID: "test-default-subnet-id",
+						},
+					},
+				},
+				service: &corev1.Service{},
+			},
+			want: "test-default-subnet-id",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := tt.args.lbaasV2.getSubnetID(tt.args.service, tt.args.svcConf)
+			if tt.expectedErr != "" || err != nil {
```

Comment:
This would indeed work in all likely cases. However, it is not quite correct.

`assert.EqualError` is designed for cases where an error is expected, and only match its message against a string.

Now consider this edge case:
```Go
func TestAssertEqualError(t *testing.T) {
	t.Run("edge case: non-nil error with empty message", func(t *testing.T) {
		var (
			err         error  = fmt.Errorf("")
			expectedErr string = ""
		)
		assert.EqualError(t, err, expectedErr) // PASS: the error message matches the expected string
	})
}
```
In this example the error is not nil, however the assertion would not trigger any test failure.

Were the function under test to return a non-nil error with an empty string as its message, this test would not catch the issue.

Clearly, we are unlikely to hit this particular case. However, by the law of large numbers, an issue like this WILL happen, and those are precisely the hardest to debug. Let's be a little defensive in our tests, at the price of a couple more lines of code.

Here you probably should write a separate block for when `tt.expectedErr` is the empty string, because in that case we really want to `assert.NoError(t, err)`.

Reply:
wow!! nice catch i really didn't consider that edge case! thanks a lot for the insight, I'll fix it